Dynamic RBAC.adoc

```diff
@@ -0,0 +1,320 @@
+Delegate storage of the mappings to a new SPI (`RolePermissionMapper`).
+Add a default `ClusterPermissionMapper` which uses a persistent internal replicated cache `org.infinispan.PERMISSIONS`
+to store the role/permission mappings. Allow usage of a custom implementation of the SPI.
+
+== Compatibility impact
+
+Ensure that statically defined role/permission mappings cannot be manipulated.
+
+== Security impact
+
+Only users with `ADMIN` permissions are allowed to interact with the API.
+
+== Configuration Schema
+
+[source,xml]
+----
+<xs:element name="authorization" minOccurs="0">
+  <xs:complexType>
+    <xs:sequence>
+      <xs:choice minOccurs="0" maxOccurs="1">
+        <xs:element name="cluster-permission-mapper">
+          <xs:annotation>
+            <xs:documentation>
+              Uses the cluster permission mapper which stores the role to permission mappings within the cluster registry.
+            </xs:documentation>
+          </xs:annotation>
+        </xs:element>
+        <xs:element name="custom-permission-mapper">
+          <xs:annotation>
+            <xs:documentation>
+              Uses a custom permission mapper.
+            </xs:documentation>
+          </xs:annotation>
+          <xs:complexType>
+            <xs:attribute name="class" type="xs:string">
+              <xs:annotation>
+                <xs:documentation>
+                  Class of the custom role to permission mapper.
+                </xs:documentation>
+              </xs:annotation>
+            </xs:attribute>
+          </xs:complexType>
+        </xs:element>
+      </xs:choice>
+    </xs:sequence>
+  </xs:complexType>
+</xs:element>
+----
+
+
+== Public API
+
+[source,java]
+----
+package org.infinispan.security;
+
+public interface RolePermissionMapper {
+   /**
+    * Sets the context for this {@link RolePermissionMapper}
+    *
+    * @param context
+    */
+   void setContext(AuthorizationMapperContext context);
+
+   /**
+    * Required permissions: <tt>ADMIN</tt>
+    * @param name the name of the role
+    * @return the {@link Role}
+    */
+   CompletionStage<Role> getRole(String name);
+
+   /**
+    * Required permissions: <tt>ADMIN</tt>
+    * @return all roles handled by this RolePermissionMapper
+    */
+   CompletionStage<Map<String, Role>> getAllRoles();
+}
+----
+
+[source,java]
+----
+package org.infinispan.security;
+
+public interface MutableRolePermissionMapper extends RolePermissionMapper {
+   /**
+    * Required permissions: <tt>ADMIN</tt>
+    */
+   CompletionStage<Void> addRole(Role role);
+
+   /**
+    * Required permissions: <tt>ADMIN</tt>
+    */
+   CompletionStage<Void> removeRole(String role);
+}
+----
+
+Introduce an `org.infinispan.security.AuthorizationMapperContext` interface that supersedes `org.infinispan.security.PrincipalRoleMapperContext`
+
+== Deprecations
+
+* `org.infinispan.security.PrincipalRoleMapperContext`
+
+== Hot Rod API
+
+N/A
```

Comment:
we don't want to allow role manipulation on hotrod? I'm thinking about the fact that we are able to configure some security or configure caches in startup. Maybe an API for authz manipulation via hotrod is something we might want to consider at a point?

Reply:
Definitely not. Defining role is a global server admin operation.